### DISCUSSION TITLE
Implementing Lazy Loading for Carousel Slides – Guidance Needed

### DISCUSSION THREAD

**Discussion by m-elagamy** [davidjerleke/embla-carousel#1058] [Help] [answered]
**Hi Embla community,**

I'm working on a project using the Shadcn Embla Carousel, and I'm interested in implementing lazy loading for slides that aren't in the viewport to optimize performance. My current setup displays 12 items (representing hourly data in a weather app), and I've noticed that even with all items pre-rendered, the app runs smoothly. However, I want to showcase my skills in performance optimization by incorporating lazy loading to only render slides when they come into view.

I've explored using slidesInView() and slidesNotInView() provided by the Embla API but ran into some challenges:

I attempted to manage state for rendering slides based on visibility, but initialSlidesInView logged as an empty array [] when checking at first.

Despite various changes, I ended up with scenarios where either all slides were rendered at once or nothing displayed at all upon page load.

**Questions:**

1. What would be the best way to implement lazy loading with Embla's API?

2. Are there recommended practices for detecting when a slide enters the viewport and triggering a load/render for it?

3. Is there an example or plugin that aligns with this use case?

Any advice, examples, or pointers to relevant documentation would be greatly appreciated!

Thanks in advance for your help!

```jsx
"use client";

import { useCallback, useEffect, useMemo, useRef, useState } from "react";
import {
  Carousel,
  CarouselContent,
  CarouselItem,
  type CarouselApi,
} from "@/components/ui/carousel";
import HourCard from "./HourCard";
import type { HourData } from "@/types/WeatherFlags";

const CarouselHours = ({ hoursToDisplay }: { hoursToDisplay: HourData[] }) => {
  const [api, setApi] = useState<CarouselApi | null>(null);
  const [scrollProgress, setScrollProgress] = useState(0);
  const animationFrame = useRef<number | null>(null);
  const [visibleHours, setVisibleHours] = useState<number[]>([]);

  const handleScroll = useCallback(() => {
    if (!api) return;

    if (animationFrame.current) cancelAnimationFrame(animationFrame.current);

    animationFrame.current = requestAnimationFrame(() => {
      setScrollProgress(api.scrollProgress());
      setVisibleHours(api.slidesInView());
    });
  }, [api]);

  useEffect(() => {
    if (!api) return;

    api.on("scroll", handleScroll);
    setVisibleHours(api.slidesInView());

    return () => {
      api.off("scroll", handleScroll);
      if (animationFrame.current) cancelAnimationFrame(animationFrame.current);
    };
  }, [api, handleScroll]);

  const memoizedHoursToDisplay = useMemo(() => {
    return hoursToDisplay;
  }, [hoursToDisplay]);

  return (
    <>
      <Carousel
        className="cursor-grab active:cursor-grabbing"
        setApi={setApi}
        opts={{ align: "start" }}
      >
        <CarouselContent>
          {memoizedHoursToDisplay?.map((hour, index) => (
            <CarouselItem
              key={hour.time}
              className="basis-1/3 lg:basis-1/4 2xl:basis-1/5"
            >
              {visibleHours.includes(index) && <HourCard hour={hour} />}
            </CarouselItem>
          ))}
        </CarouselContent>
      </Carousel>
      <div
        className="absolute -left-full bottom-0 h-[3px] w-full rounded-sm bg-gradient-to-r from-slate-200/30 to-slate-400/50 transition-transform duration-500 ease-out"
        style={{
          transform: `translate3d(${scrollProgress * 100}%,0,0)`,
        }}
      />
    </>
  );
};

export default CarouselHours;
```

**Comment by davidjerleke** [comment, marked as answer]
Hi @Mahmoud-Elagamy,
Thank you for your questions.

> What would be the best way to implement lazy loading with Embla's API?

I would advise against updating slides in view during scroll because that's not how it's designed to be used. There's a much more performant way of doing this:

```tsx
  const [visibleHours, setVisibleHours] = useState<number[]>([]);

  useEffect(() => {
    if (!api) return;

    const updateSlidesInView = (api: CarouselApi) => {
      setVisibleHours(api.slidesInView());
    };

    api.on("slidesInView", updateSlidesInView);
    updateSlidesInView(api);

    return () => {
      api.off("slidesInView", updateSlidesInView);
    };
  }, [api]);
```

> Are there recommended practices for detecting when a slide enters the viewport and triggering a load/render for it?

Generally, I would recommend that you have empty slides that you populate with content as they enter the viewport. This of course, is based on the assumption that you know how many total slides there will be in advance. If this isn't possible for your use case, I would advise you to use the [infinite scroll example approach](https://www.embla-carousel.com/examples/predefined/#infinite-scroll) in the docs. 

> Is there an example or plugin that aligns with this use case?

At the time of writing, there's no plugin that handles these cases out of the box.

**Comment by m-elagamy** [comment]
Thank you so much for your previous response and guidance on implementing lazy loading in my project! I appreciate the time and detail you put into helping me.
